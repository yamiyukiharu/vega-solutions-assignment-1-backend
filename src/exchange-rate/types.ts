import { Currency } from "src/common/enums";

export type HistoricalDataResult = {
  time: Date;
  value: string;
};

export type GetHistoricalRateOptions = {
  from: Currency;
  to: Currency;
  start?: Date;
  end?: Date;
};