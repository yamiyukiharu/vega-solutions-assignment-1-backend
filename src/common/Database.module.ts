import { Module } from "@nestjs/common";
import { TypeOrmModule } from "@nestjs/typeorm";
import { AppConfigModule } from "./AppConfig.module";
import { ConfigService } from "@nestjs/config";
import { Transaction } from "src/transaction/models/Transaction.model";

@Module({
  imports: [
    TypeOrmModule.forRootAsync({
      imports: [AppConfigModule],
      useFactory: (configService: ConfigService) => ({
        type: 'mongodb',
        url: configService.get('DB_URL'),
        entities: [Transaction],
        synchronize: true,
        useNewUrlParser: true,
        logging: true,
      }),
      inject: [ConfigService],
    }),
  ],
  exports: [TypeOrmModule]
})
export class DatabaseModule { }