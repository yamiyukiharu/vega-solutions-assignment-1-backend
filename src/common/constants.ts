export const REPORTS_QUEUE = 'transactions-report';

export const RECORD_QUEUE = 'transactions-record';