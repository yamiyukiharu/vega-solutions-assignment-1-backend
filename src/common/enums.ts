export enum Protocol {
  UNISWAPV3 = 'uniswapv3',
}

export enum Pool {
  ETH_USDC = 'usdc_eth',
}

export enum ReportStatus {
  PENDING = 'pending',
  IN_PROGRESS = 'in_progress',
  COMPLETED = 'completed',
  FAILED = 'failed',
}