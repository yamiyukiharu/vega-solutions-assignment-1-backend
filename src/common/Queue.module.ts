
import { BullModule } from '@nestjs/bull';
import { Module } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { AppConfigModule } from './AppConfig.module';

@Module({
  imports: [
    BullModule.registerQueueAsync({
      imports: [AppConfigModule],
      name: 'transactions-report',
      useFactory: async (configService: ConfigService) => ({
        redis: {
          host: configService.get('QUEUE_HOST'),
          port: configService.get('QUEUE_PORT'),
        },
      }),
      inject: [ConfigService],
    }),
  ],
  exports: [BullModule],
})
export class QueueModule { }
