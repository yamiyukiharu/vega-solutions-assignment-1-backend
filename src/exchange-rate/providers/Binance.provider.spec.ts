import { Test, TestingModule } from '@nestjs/testing';
import { HttpModule, HttpService } from '@nestjs/axios';
import { BinanceProvider } from './Binance.provider';
import { GetHistoricalRateOptions } from '../types';
import { AxiosResponse } from 'axios';
import { of } from 'rxjs';
import { Currency } from 'src/common/enums';

describe('BinanceProvider', () => {
  let provider: BinanceProvider;
  const mockAxiosGet = jest.fn();

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        BinanceProvider,
        {
          provide: HttpService,
          useValue: {
            axiosRef: {
              get: mockAxiosGet,
            },
          },
        },
      ],
    }).compile();

    provider = module.get<BinanceProvider>(BinanceProvider);
  });

  describe('convert', () => {
    it('should return the exchange rate between two currencies', async () => {
      const from = Currency.ETH;
      const to = Currency.USDT;
      const symbol = 'ETHUSDT';
      const price = 50000;

      mockAxiosGet.mockResolvedValue({
        data: {
          price,
        },
      });

      const result = await provider.convert(from, to);

      expect(mockAxiosGet.mock.calls[0]).toMatchSnapshot();
    });
  });

  describe('getHistoricalRates', () => {
    it('should return historical exchange rates between two currencies', async () => {
      const from = Currency.ETH;
      const to = Currency.USDT;
      const symbol = 'ETHUSDT';
      const startTimestamp = 1614556800; // March 1, 2021 12:00:00 AM GMT
      const endTimestamp = 1614643199; // March 1, 2021 11:59:59 PM GMT
      const data = [
        [
          1614556800000,
          '47000.00000000',
          '48000.00000000',
          '46000.00000000',
          '47000.00000000',
          '1000.00000000',
          1614556859999,
          '47000000.00000000',
          500,
          '500.00000000',
          '100000000.00000000',
          '0',
        ],
        [
          1614556860000,
          '48000.00000000',
          '49000.00000000',
          '47000.00000000',
          '48000.00000000',
          '2000.00000000',
          1614556919999,
          '96000000.00000000',
          1000,
          '1000.00000000',
          '100000000.00000000',
          '0',
        ],
        [
          1614556920000,
          '48000.00000000',
          '49000.00000000',
          '47000.00000000',
          '48000.00000000',
          '3000.00000000',
          1614556979999,
          '144000000.00000000',
          1500,
          '1500.00000000',
          '100000000.00000000',
          '0',
        ],
      ];

      mockAxiosGet.mockResolvedValue({
        data,
      });

      const result = await provider.getHistoricalRates({
        from,
        to,
        startTimestamp,
        endTimestamp,
      });

      expect(result).toEqual([
        {
          timestamp: 1614556800,
          value: '47000.00000000',
        },
        {
          timestamp: 1614556860,
          value: '48000.00000000',
        },
        {
          timestamp: 1614556920,
          value: '48000.00000000',
        },
      ]);
      
      expect(mockAxiosGet.mock.calls[0]).toMatchSnapshot();

    });
  });
});
