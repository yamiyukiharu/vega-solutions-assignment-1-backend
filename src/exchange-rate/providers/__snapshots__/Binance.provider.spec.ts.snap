// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BinanceProvider convert should return the exchange rate between two currencies 1`] = `
[
  "https://data.binance.com/api/v3/ticker/price?symbol=ETHUSDT",
]
`;

exports[`BinanceProvider getHistoricalRates should return historical exchange rates between two currencies 1`] = `
[
  "https://data.binance.com/api/v3/ticker/price?symbol=ETHUSDT",
]
`;
